### ISSUE TITLE
Best practices for common patterns

### ISSUE THREAD

**Issue by Spiralis** [schyntax/schyntax#14]
I am compiling some best practices for common scenarios for my users. Can anyone help make sure that these are as simple and intuitive as possible, while still being accurate?

| Trigger once every ... | Expression(s) |
| --- | --- |
| ... second | `sec(*)` |
| ... minute | `min(*)` |
| ... hour | `hour(*)` |
| ... day | `day(*)`, `date(*)`, `dow(*)`, `dom(*)` |
| ... week | `dow(1)` |
| ... month | `dom(1)` |
| ... year | `date(1/1)` |

| Trigger on the ... | first | n-th | last |
| --- | --- | --- | --- |
| ... day of week: | `dow(1)` | `dow(n)` | `dow(7)` |
| ... day of month: | `dom(1)` | `dom(n)` | `dom(-1)` |
| ... day of year: | `date(1/1)` | ? | `date(12/31)` |
| ... week of month: | `dom(1)` | ? | `dom(-7)` |
| ... week of year: | `date(1/1)` | ? | `date(12/24)` |
| ... month of year: | `date(1/1)` | ? | `date(12/1)` |

**Comment by bretcope** [comment]
I generally avoid using number arguments for day of week. It's valid, but naming the day of week is better practice `dow(sun)` and less ambiguous (at least to English speakers).

**Comment by Spiralis** [comment]
As I was compiling this list I noticed some improvements that could be made. I might be missing something vital from my understanding though so please help correct me if I am wrong on anything. 
If not then I can't help but feeling like a couple of specifiers are missing. In particular the basic `week()`, `month()`, `year()` and `dayOfYear()` forms, matching the second, minute and hour patterns. 
- `week` already has another proposal (https://github.com/schyntax/js-schyntax/issues/5 - also in the roadmap). I feel that for the usage scenarios described in the proposal the name should be `weekOfMonth` and `fullWeekOfMonth` instead of `weeks`. It is more intuitive that `week` is about any week in the year, in the range from 1..52. Which also fits nicely in with the patterns from the other specifiers.  
- `month` would also be a very nice addition - and it really stands out as missing in the range. But, the same way as `w` had to be removed to make room for `week` (ref proposal https://github.com/schyntax/js-schyntax/issues/5), `m` (alias for `minute`) needs to be removed since there will then be two specifiers starting with m, making it ambigious. 
- `year` is probably an edge-case not seen often. But, it would be nice for completeness. And I do have use-cases where this would be interesting. Using `date(1/1)` works, but `year(*)` fits nicely in with the others, so - why not?
- I also want to pitch `dayOfYear`as a way to trigger on a given day of the year. The date specifier seems to be geared towards actual dates, which is fine, so this addition would be nice.  

With these implemented the _every_ patterns would be simplified to:

| Trigger once every ... | Expression(s) |
| --- | --- |
| ... second | `sec(*)` |
| ... minute | `min(*)` |
| ... hour | `hour(*)` |
| ... day | `day(*)`, `date(*)`, `dow(*)`, `dom(*)`, **[proposed]** `dayOfYear(*)` |
| ... week **[proposed]** | `week(*)` |
| ... month **[proposed]** | `month(*)` |
| ... year **[proposed]** | `year(*)` |

The _first/n-th/last_ patterns would be simplified to:

| Trigger on the ... | first | n-th | last |
| --- | --- | --- | --- |
| ... day of week: | `dow(1)` | `dow(n)` | `dow(-1)` **(1)** |
| ... day of month: | `dom(1)` | `dom(n)` | `dom(-1)` |
| ... day of year **[proposed]** | `dayOfYear(1)` | `dayOfYear(n)` | `date(-1)` **(2)** |
| ... week of month **[proposed]** | `weekOfMonth(1)` | `weekOfMonth(n)` | `weekOfMonth(-1)` |
| ... fullWeek of month **[proposed]** | `fullWeekOfMonth(1)` | `fullWeekOfMonth(n)` | `fullWeekOfMonth(-1)` **(3)** |
| ... week of year **[proposed]** | `week(1)` | `week(n)` | `week(-1)` |
| ... fullWeek of year **[proposed]** | `fullWeek(1)` | `fullWeek(n)` | `fullWeek(-1)`  **(3)** |
| ... month of year **[proposed]** | `month(1)` | `month(n)` | `month(-1)` |

**(1)**: The dayOfWeek does not accept negative values. Any particular reason for this? It would be nice if all the basic specifiers worked the same way. So, I propose that dayOfWeek also understands negative values, for completeness.

**(2)**: If all specifiers should match really well then day should have been enumerated in regards to year, as in day(1..365). But, that will introduce a notable breaking change. 

**(3)**: I added fullWeekOfMonth and fullWeek as its counterparts were proposed in https://github.com/schyntax/js-schyntax/issues/5.

**Comment by Spiralis** [comment]
@bretcope I agree on the best practices for naming days. I guess, in regards to my proposal above, month-names should be valid for the `month` specifier too.

**Comment by bretcope** [comment]
Wow, thanks for the interest. Hopefully you're equally interested in contributing some code... :)

Day of week does not accept negative values because there's no reason for it to. Every week has the same number of days. The only reason day of month accepts it is compensate for the fact that different months have different days. I'm really scratching my head to see why you would ever want to use a negative value. It feels like it's more like to confuse people than help them, especially given that there's already a better way of specifying day of week (e.g. `wednesday`). It obviously makes sense to support negative values for week expressions, since they have the same problem as day of month.

I don't agree about `week()` as a week-of-year expression being more intuitive or more consistent with the other expressions. Everything else is relative to the next biggest unit (minute is relative to hours, not days; the `day` alias is for day of week, not day of month, etc). I'm also not convinced that week of year is a more common use case than week of month.

However, I do like the idea of adding both `weekOfMonth` and `weekOfYear`, and I don't have any opposition to `dayOfYear`.

I have mixed feelings about `year`. I could see it being useful inside a group, but it's typical usefulness would be rather low. Ultimately I think it would be fine to add it. Something that probably needs to be addressed in the current Schyntax implementations is that they only search up to one year in the future or past for an event date. That was a hack to prevent an infinite loop in case there was no valid time, and the very original implementation didn't allow years to be specified in the date expression. The algorithm could probably be a little more intelligent about how far it should try to search based on the schedule.

---

If you're interested in contributing, I would pick one, implement it, and submit it as a PR to the C# implementation. I would pick `dayOfYear` since it's the simplest. Make sure to [submit tests](https://github.com/schyntax/schyntax/blob/master/WritingTests.md) to this repo.

**Comment by bretcope** [comment]
Oh, and I have no problem with `month()` or named month literals (e.g. `feb` or `february`) either. I think that's much more useful than year.

**Comment by Spiralis** [comment]
Cool. I am fine with all your comments actually. 
- The -1 for days was mostly for completeness and consistency (as other numeric specifiers have them). 
- As for weeks I agree that most scenarios are probably related to usage in months. 
- year is an edge case for scenarios. I do have them, but if - like you said there are limitations in how far ahead they search for dates then this might have to be considered as a part of such a fix. It would be nice for completeness though.

I'd love to contribute. I have an oncoming deadline though, so I cannot guarantee how much I will be able to contribute on a short term, but I'll do my best :)

**Comment by Spiralis** [comment]
I have completed most of what I can contribute on in regards to `dayOfYear`, with PRs:
- https://github.com/schyntax/schyntax/pull/15
- https://github.com/schyntax/cs-schyntax/pull/4
- https://github.com/schyntax/js-schyntax/pull/11

I have also created a PR for `month`, but on this one I have a failing test that I require assistance on:
- https://github.com/schyntax/cs-schyntax/pull/3

In regards to this, maybe the `m` abbreviation should be deprecated (like the `w`) as it will be ambiguous with `month` added. `m` still means `min` in the month-PR, but it might be better to disallow it?

Finally I would love to see `week` and `weekOfYear` added. But, any contribs I could make to that will have to come later.
